### ISSUE TITLE
Metrics preprocessing report quality validation numbers for reporting

### ISSUE THREAD

**Issue by awig** [SolarArbiter/solarforecastarbiter-core#299]
Calculate total number of set quality flags in the metrics preprocessing on the provided `Observation`.

Update the `get_validation_issues` [function](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/20cee821dd0c872d6878fe3d7d5e92d9e23e4cd6/solarforecastarbiter/reports/main.py#L142 ) using a dictionary with actual calculated values.

Questions:
* Report for each quality flag type individually (not counting the intersection)?
* I think we should also add a total quality flag that takes the union of all quality flags that are to be active, so that way the user can easily understand the amount of "bad" data.
* Do we also need to report the number of NaNs from the resampling and aligning? #290

**Comment by wholmgren** [comment]
> Report for each quality flag type individually (not counting the intersection)?

Report the number of points flagged for each type individually (like in the dummy data). 

> I think we should also add a total quality flag that takes the union of all quality flags that are to be active, so that way the user can easily understand the amount of "bad" data.

Sounds good

> Do we also need to report the number of NaNs from the resampling and aligning? #290

Yes, though I don't know if you want to take it on all together or separately.

**Comment by awig** [comment]
> Yes, though I don't know if you want to take it on all together or separately.

Let's do it separately for now.

**Comment by alorenzo175** [comment]
The simple counting of quality flags should be taken care of in the report refactor https://github.com/SolarArbiter/solarforecastarbiter-core/pull/291/files#diff-715b43f6fac14c48d8729da1c197d286R52. This still has to be incorporated into the ProcessedForecastObservation and then make it's way to the final report

**Comment by awig** [comment]
@alorenzo175 OK. I see it. Should I merge with your branch then to work on the calculating each individual quality flag counts?

**Comment by awig** [comment]
Per today's call. I'll merge and continue on working when the new beta4 branch is created.

**Comment by awig** [comment]
For now, I'm planning on just reporting quality flag counts in the following type of table:

|                     |  **Totals**       | **<qflag1-name>** | **<qflag2-name>** |  ...                        |
| ------------ | -------------- | ------------------ | --------------------- | ---------------- | 
| **Count**    | 2000               |  200                       | 20                               |  ...                        |
| **Percent** | 100%               | 10%                        | 1%                               | ...                        |

David and I have also discussed other graphical representations that we could explore in the future:

![20200109_143125](https://user-images.githubusercontent.com/6306150/72110423-0c1bae00-32ed-11ea-8cb8-8b3f89082a48.jpg)

**Comment by alorenzo175** [comment]
1.0beta4 branch now has what's refactored for reports so far

**Comment by awig** [comment]
Great thanks. I'll merge and continue work.

**Comment by awig** [comment]
@alorenzo175 Seems like you've added thing for recording all the quality flag counts at least as `ValidationResult`s. (so my job here is done 😛 ) 

So, it seems I just need to do the following:

1. add these quality flag counts to a table in the reports (i didn't see anything explicitly done yet)
2. add a count for any dropped nans before the alignment of the observation (name: UNAVAILABLE)
3. add a count for the total validation+dropped observations (name: TOTAL or FINAL)
4. add these to item 2,3 counts into table of item 1

Anything else I'm missing?

**Comment by alorenzo175** [comment]
I think that all sounds right

**Comment by awig** [comment]
These are the additional counts that I ended up adding:

* **Observations Unavailable** = (after apply validation) count all nans using `isna().sum()`
* **Observations Removed from Resampling and Alignment** = length of observations after apply validation - length of observations after resampling and alignment
* **Forecasts Removed from Alignment** = length of raw forecasts - length of forecasts after resampling and alignment
* **Total Observations Removed** = length of raw observations - length of observations after resampling and alignment

Here is an example of the list of validation results:

```python
raw_report.processed_forecasts_observations[0].validation_results

(ValidationResult(flag='NIGHTTIME', count=23873),
 ValidationResult(flag='USER FLAGGED', count=5),
 ValidationResult(flag='INTERPOLATED VALUES', count=0),
 ValidationResult(flag='LIMITS EXCEEDED', count=355),
 ValidationResult(flag='INCONSISTENT IRRADIANCE COMPONENTS', count=0),
 ValidationResult(flag='STALE VALUES', count=0),
 ValidationResult(flag='Observations Unavailable', count=0),
 ValidationResult(flag='Observations Removed from Resampling and Alignment', count=17622),
 ValidationResult(flag='Forecasts Removed from Alignment', count=689),
 ValidationResult(flag='Total Observations Removed', count=41500))
```

**Comment by awig** [comment]
I think maybe I should use "Ignored" instead of "Removed".

**Comment by alorenzo175** [comment]
What about when the forecast and observation do not have the same interval length? Then I don't think using Ignored in Observations Ignored from Resampling and Alignment and Total Observations Ignored is quite right since some of those observations are included in the mean calculation. 

Instead, maybe there should be something like TOTAL FLAGGED VALUES = length before validation - length after excluding. And Observations Ignored from Resampling and Alignment could then just be length obs_resampled before dropping/alignment - length final obs_resampled

**Comment by awig** [comment]
True. I'll update them. Thanks

**Comment by awig** [comment]
I'm using "TOTAL FLAGGED VALUES DISCARDED" to distinguish between flags that are filtered and are not.

Also, `ValidationResult` could potentially be misleading for the resampling and alignment process. Should I create a separate `ResampleAlignResult` object or continue to use `ValidationResult`?

**Comment by alorenzo175** [comment]
> I'm using "TOTAL FLAGGED VALUES DISCARDED" to distinguish between flags that are filtered and are not.

:+1: 

> Also, `ValidationResult` could potentially be misleading for the resampling and alignment process. Should I create a separate `ResampleAlignResult` object or continue to use `ValidationResult`?
 
I agree putting it all in ValidationResult could be misleading. A separate object is a good idea.

**Comment by awig** [comment]
Thanks. How about `PreprocessingResult`?  I'm structuring in such away that it is very similar to ValidationResult but with a `name` instead of `flag` parameter.  Adding the actual calculations to `metrics.resample_and_align` except for the TOTAL FLAGGED VALUES DISCARDED which remains in `process_forecast_observations`

**Comment by awig** [comment]
@alorenzo175 I'm done with the `PreprocessingResult` and adding it to the report (all tests are passing) and have started working on adding it to the html rendering. I was wondering why you use the an html file for the Observations/Forecast table in obsfx_table.html instead of adding a bokeh DataTable widget? Is it needed for links?  I'm working on using the bokeh tables, but please let me know if i should not.

**Comment by alorenzo175** [comment]
We had some issues with how the bokeh DataTable was rendering things compared to CSS on HTML tables, and I don't really see how linking between the table and any of the plots is beneficial. I think it's also more natural later to make Latex tables for PDF output instead of rendering the bokeh table as SVG.

**Comment by alorenzo175** [comment]
So I would say just make an HTML table using jinja

**Comment by wholmgren** [comment]
I didn't realize the bokeh table was being replaced but I am all for it. It's ugly and has been a major pain to debug. The only benefit was the sorting feature, but we could probably find a simple javascript implementation for that.

**Comment by awig** [comment]
Thanks for the quick reply! I won't use bokeh DataTable.  I was also going to put the Validation counts into a similar table.

**Comment by alorenzo175** [comment]
closed in #321
